File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        unordered_set<ListNode*> st;

        // Store nodes of list A
        ListNode* temp = headA;
        while (temp != NULL) {
            st.insert(temp);
            temp = temp->next;
        }

        // Check nodes of list B
        temp = headB;
        while (temp != NULL) {
            if (st.find(temp) != st.end()) {
                return temp;   // Intersection found
            }
            temp = temp->next;
        }

        return NULL; // No intersection
    }
};

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode temp = headA; 
        Map<ListNode , Integer>mp = new HashMap<>(); 
        while(temp!= null){
            mp.put(temp,1);
            temp = temp.next;
        }
        temp = headB; 
        while(temp!=null){
            if(mp.containsKey(temp)) return temp; 
            temp = temp.next; 
        }
        return null;
    }
}